include/bluetooth/bluetooth.h

```diff
@@ -398,6 +475,134 @@ int bt_le_adv_update_data(const struct bt_data *ad, size_t ad_len,
  */
 int bt_le_adv_stop(void);
 
+/** @brief Increment an advertising set`s reference count.
+ *
+ *  Increment the reference count of an advertising set object.
+ *
+ *  @param adv Advertising set object.
+ *
+ *  @return Advertising set object with incremented reference count.
+ */
+struct bt_adv *bt_adv_ref(struct bt_adv *adv);
+
+/** @brief Decrement an advertising set's reference count.
+ *
+ *  Decrement the reference count of an advertising setobject.
+ *
+ *  @param adv Advertising set object.
+ */
+void bt_adv_unref(struct bt_adv *adv);
+
+/** @brief Create advertising set.
+ *
+ *  Create a new advertising set and set advertising parameters.
+ *
+ *  @param param Advertising parameters.
+ *
+ *  @return Zero on success or (negative) error code otherwise.
+ *  @return -ECONNREFUSED When connectable advertising is requested and there
+ *			  is already maximum number of connections established.
+ *			  This error code is only guaranteed when using Zephyr
+ *			  controller, for other controllers code returned in
+ *			  this case may be -EIO.
+ */
+struct bt_adv *bt_le_adv_set_create(const struct bt_le_adv_param *param);
+
+/** Start advertising with the given advertising set
+ *
+ *  Start advertising with a specific advertising set.
+ *  If both timeout and advertising event limit is given then the advertising
+ *  set will be stopped once one of them is reached.
+ *
+ *  @param adv         Advertising set object.
+ *  @param timeout     Advertiser timeout (N * 10 ms), Application will be
+ *                     notified by the advertiser sent callback. Set to zero
+ *                     for no timeout.
+ *  @param num_events  Number of advertising events. Application will be
+ *                     notified by the advertiser sent callback. If used
+ *                     together with timeout then whichever happens first will
+ *                     stop the advertiser. Set to zero for no limit.
+ */
+int bt_le_adv_set_start(struct bt_adv *adv, u16_t timeout, u8_t num_events);
```

Comment:
Same here. The object operated on should match `bt_le_<object>_start()`

Reply:
The problem is that bl_le_adv_start already exists

Reply:
Right. That's a pity. I think we might need to go for the other option then, i.e. rename `bt_adv` to `bt_adv_set`. Actually, there's another inconsistency: the prefix for the functions is `bt_le_adv_*` but you've omitted "le" from the type. We have "le" in other advertising-related types already, like `bt_le_adv_param` so it'd be good to try to keep consistency. (If I were to were to start from scratch I might consider omitting it everywhere since BR/EDR doesn't have advertising, but here we are :)

Reply:
Btw, I was almost included to suggest we emit "le" for all the new stuff, but then it'd make things really confusing to have both `bt_adv_start()` and `bt_le_adv_start()`. One other thought: is "start" really the best verb? Is there something else that'd make equal (or more) sense?

Reply:
Using `bt_le_adv_set` as the prefix everywhere becomes quite long. And also I'm not really happy with using `set` in the name. Advertising Set is the HCI name, and it never really made much sense to me. Also set is confusing as it is mostly related to get/set functions.

At the moment `bt_le_adv_param` is re-used so it creates an inconsistency. I agree, but maybe that is the best option anyway.

Using `bt_adv` as prefix:
```
struct bt_adv_cb {
void bt_adv_cb_register(struct bt_adv_cb *cb);
struct bt_adv *bt_adv_create(const struct bt_le_adv_param *param);
int bt_adv_start(struct bt_adv *adv, u16_t timeout, u8_t num_events);
int bt_adv_stop(struct bt_adv *adv);
int bt_adv_data(struct bt_adv *adv,
int bt_adv_delete(struct bt_adv *adv);
u8_t bt_adv_get_index(struct bt_adv *adv);
int bt_adv_get_info(const struct bt_adv *adv, struct bt_adv_info *info);
```
New scan listener callback API which is not released, here we adde `bt_le_scan`,
maybe we could undo that?
```
typedef void bt_scan_cb_t(const bt_addr_le_t *addr, s8_t rssi,
			  u8_t adv_type, struct net_buf_simple *buf);
struct bt_adv_info {
}

struct bt_scan_cb {
	void (*recv)(const struct bt_adv_info *info,
 		     struct net_buf_simple *buf);
void bt_scan_cb_register(struct bt_le_scan_cb *cb);
```

Also since we have in a way made the `cb` parameter in bt_le_scan_start a bit obsolete
we could consider to add `bt_scan_start` that removes this argument

Using start/stop I think makes sense, at least I cannot find a better name that would
justify changing it from what the user expects.
We also use start/stop in nRF5 SDK and SoftDevice

Reply:
In general I agree with your line of thought here. One problem with "scan" is that it's also a BR/EDR concept. You enable inquiry scan to be discoverable and page scan to be connectable. So it might be confusing to omit "le". Remember: we considered removing BR/EDR but were requested to keep it there since at least one member company had plans to make a dual-mode product.

Reply:
..having said that, we don't actually expose inquiry & page scan in our API, instead we have `bt_br_set_discoverable()` and `bt_br_set_connectable()`. So perhaps `bt_scan_*` would be ok.

Reply:
@trond-snekvik would appreciate your thoughts here as well on the naming

Reply:
I would argue that `bt_adv` as an object is simply not accurate nor informative. What we are talking about here are advertising sets, just like we have connections. Hence my proposal is to:
- remove `_le_`
- Use `adv_set` consistently throughout the whole API

Reply:
Is `adv_set` really the right term though, considering that it needs to cover legacy 4.0 advertising? IMO, @joerchan's suggestion makes the most sense, an advertiser sounds like an instantiable object without breaking with the pre-AE language or causing confusion in AE.

I'm not sure I fully agree with removing `_le_` though, as we'd be stepping on a namespace we might want to add support for later. If we do, I think it should be unified across the public facing API eventually, there are a couple of other instances of `_le_` usage around:
- `bt_addr_le_t`
- conn.h is a mix of `bt_conn_<...>` and `bt_conn_le_<...>`
- hci.h mostly uses `bt_hci_le_`
- l2cap.h uses `bt_l2cap_le_`

Reply:
> `bt_addr_le_t`

That's basically for the <type> <addr> tuple that's not just to reduce needed function call parameters but also maps in it's binary representation to how these values occur in HCI packets. Additionally, there are commands/events valid for LE with only the 6-byte address part, whenever the type is implicit, so we still need a bt_addr_t. So I'm not too concerned for keeping this around

> hci.h mostly uses bt_hci_le_

That's just a direct mapping to the HCI spec. The LE commands are called HCI_LE_<something>. So I think this is rather independent from the naming discussion.

The other two bullet points are valid for consideration though

Reply:
After the meeting, agreed to:

```

bt_le_ext_adv_create(void, params)
bt_le_ext_adv_delete(bt_le_ext_adv*)
bt_le_ext_adv_update_params(bt_le_ext_adv*, params)
bt_le_ext_adv_set_data(bt_le_ext_adv*, data)

bt_le_ext_adv_start(bt_le_ext_adv*, timeout )
bt_le_ext_adv_stop(bt_le_ext_adv* )
```